plugins/cuda/cuda_plugin.c

```diff
@@ -0,0 +1,406 @@
+	}
+interrupt:
+	int_ret = interrupt_restore_thread(restore_pid);
+
+	return ret != 0 ? ret : int_ret;
+}
+CR_PLUGIN_REGISTER_HOOK(CR_PLUGIN_HOOK__CHECKPOINT_DEVICES, cuda_plugin_checkpoint_devices);
+
+int cuda_plugin_pause_devices(int pid)
+{
+	int restore_pid;
+	char msg_buf[64];
+
+	if (plugin_disabled) {
+		return 0;
+	}
+
+	restore_pid = get_cuda_restore_pid(pid);
+
+	if (restore_pid == -1) {
+		pr_info("no need to pause devices on pid %d\n", pid);
+		return 0;
+	}
+
+	pr_info("pausing devices on pid %d\n", pid);
+	int status = cuda_process_checkpoint_action(pid, ACTION_LOCK, opts.timeout * 1000, msg_buf, sizeof(msg_buf));
+	if (status) {
+		pr_err("PAUSE_DEVICES failed with %s\n", msg_buf);
+		return -1;
+	}
+	if (add_pid_to_buf(&cuda_pids, pid)) {
+		pr_err("unable to track paused pid %d\n", pid);
+		status = cuda_process_checkpoint_action(pid, ACTION_UNLOCK, 0, msg_buf, sizeof(msg_buf));
+		if (status) {
+			pr_err("Failed to unlock process status %s, pid %d may hang\n", msg_buf, pid);
+		}
+		return -1;
+	}
+
+	return 0;
+}
+CR_PLUGIN_REGISTER_HOOK(CR_PLUGIN_HOOK__PAUSE_DEVICES, cuda_plugin_pause_devices)
+
+int resume_device(int pid, int checkpointed)
+{
+	char msg_buf[CUDA_CKPT_BUF_SIZE];
+	int status;
+	int ret = 0;
+	int int_ret;
+
+	int restore_pid = get_cuda_restore_pid(pid);
+	if (restore_pid == -1) {
+		pr_info("No need to resume devices on pid %d\n", pid);
+		return 0;
+	}
+
+	pr_info("resuming devices on pid %d\n", pid);
+	/* The resuming process has to stay frozen during this time otherwise
+	 * attempting to access a UVM pointer will crash if we haven't restored the
+	 * underlying mappings yet
+	 */
+	pr_debug("Restore thread pid %d found for real pid %d\n", restore_pid, pid);
+	/* wakeup the restore thread so we can handle the restore for this pid,
+	 * rseq_cs has to be restored before execution
+	 */
+	if (ptrace(PTRACE_CONT, restore_pid, NULL, 0)) {
+		pr_err("Unable to resume cuda restore pid %d on root pid %d\n", restore_pid, pid);
+		return -1;
+	}
+
+	if (checkpointed) {
+		status = cuda_process_checkpoint_action(pid, ACTION_RESTORE, 0, msg_buf, sizeof(msg_buf));
+		if (status) {
+			pr_err("RESUME_DEVICES RESTORE failed with %s\n", msg_buf);
+			ret = -1;
+			goto interrupt;
+		}
+	}
+
+	status = cuda_process_checkpoint_action(pid, ACTION_UNLOCK, 0, msg_buf, sizeof(msg_buf));
+	if (status) {
+		pr_err("RESUME_DEVICES UNLOCK failed with %s\n", msg_buf);
+		ret = -1;
+	}
+
+interrupt:
+	int_ret = interrupt_restore_thread(restore_pid);
+
+	return ret != 0 ? ret : int_ret;
+}
+
+int cuda_plugin_resume_devices_late(int pid)
+{
+	if (plugin_disabled) {
+		return 0;
+	}
+
+	return resume_device(pid, 1);
+}
+CR_PLUGIN_REGISTER_HOOK(CR_PLUGIN_HOOK__RESUME_DEVICES_LATE, cuda_plugin_resume_devices_late)
+
+int cuda_plugin_init(int stage)
+{
+	if (!cuda_checkpoint_supports_flag("--action")) {
+		pr_err("cuda-checkpoint --action flag not supported, an r555 or higher version driver is required. Disabling CUDA plugin\n");
```

Comment:
It probably should be pr_warn and I think we need to print it only if we have the nvidia device on the host.